src/main/java/org/elasticsearch/common/lucene/Lucene.java

```diff
@@ -724,4 +726,45 @@ public void delete() {
             throw new UnsupportedOperationException("This IndexCommit does not support deletions");
         }
     }
+
+    /**
+     * Retruns <code>true</code> iff the store contains an index that contains segments that where
+     * not upgraded to the lucene 4.x format.
+     */
+    public static boolean indexNeeds3xUpgrading(Directory directory) throws IOException {
+        try {
+            SegmentInfos si =  Lucene.readSegmentInfos(directory);
+            for (SegmentCommitInfo info : si) {
+                Version version = info.info.getVersion();
+                if (version == null || version.onOrAfter(Version.LUCENE_4_0_0_ALPHA) == false) {
+                    boolean upgraded = false;
+                    final String markerFileName = IndexFileNames.segmentFileName(info.info.name, "upgraded", Lucene3xSegmentInfoFormat.UPGRADED_SI_EXTENSION);
+                    for(String file : info.files()) {
+                        if (file.equals(markerFileName)) {
+                            upgraded = true;
+                            break;
+                        }
+                    }
+                    if (upgraded == false) {
+                        return true;
+                    }
+                }
+            }
+            return false;
+        } catch (IndexNotFoundException e) {
+            return false;
+        }
+    }
+
+    public static void upgradeLucene3xSegments(Directory directory) throws IOException {
```

Comment:
this method name might be confusing? it sounds like it will do a full upgrade, while we just need a non 3x commit point?